### DISCUSSION TITLE
Technical Improvements to enhance the usability of The Turing Way Book (GSoC 2023)

### DISCUSSION THREAD

**Discussion by arya1302** [the-turing-way/the-turing-way#3081] [Ideas]
## Summary 

Over the past four years, the book has undergone significant growth, making it challenging to navigate. The project/idea focuses on enhancing the usability of The Turing Way book by improving the user interface and creating a more personalized browsing experience based on the user's profile. This involves making some additional improvements to  an existing Python package developed by the team  in 2022 integrating the updated package with the book.

## What needs to be done?

- Making browsing experience less confusing by displaying the tag of the selected ‘pathway’, rather than all tags associated with a chapter.
- Creating a new feature like 'Description' in [profiles.yml](https://github.com/alan-turing-institute/bio-Turing-Way/blob/main/master/profiles.yml) page for each profile.

## Who can help?

Users of The Turing Way book who may have suggestions for improving the usability of the Turing Way book are encouraged to share their feedback.

**Comment by malvikasharan** [comment]
Thanks for this Arya. I have moved the questions you tested today on @Arielle-Bennett and @GabinWK, integrating their feedback, in this form: [https://forms.gle/RnycLrfH1fZ5biyF8](https://forms.gle/RnycLrfH1fZ5biyF8).  You can share them in the Slack channel this week and we can gather some more response to inform your work.

**Comment by arya1302** [comment]
## Status Update (June 15, 2023)

### Work Done :
-   Learned more about the Turing Way community during the community bonding period.
-   Interacted with a few people and received their feedback about the book’s usability.
-   Implemented a description feature for each pathway as mentioned in the proposal and sent the pull request. [link #42](https://github.com/alan-turing-institute/bio-Turing-Way/pull/42)
-   Discovered the Web Speech API, which could assist in text-to-speech conversion, and explored it briefly. [link : Web Speech API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Speech_API)

### Blockers :
   - Faced an error related to pre-commit hook while trying to push the work to GitHub, but I was able to fix it.
   
### Planned :
-   Explore more about Web Speech API ( learn and test its working).
-   Work on making the browsing experience less confusing by modifying chapter links and profile tags based on a selected 'pathway,' instead of displaying all tags associated with a chapter ( as mentioned in the proposal ).

**Comment by arya1302** [comment]
## Status Update (June 21, 2023)

### Work Done :

-   Fixed the deployment issue encountered during testing of pull request [#42](https://github.com/alan-turing-institute/bio-Turing-Way/pull/42)
-   Explored online resources, including tutorials, and articles, to gain an understanding of the implementation of Web Speech API in web applications.    
-   Worked on the issue of displaying all profile tag/labels at the top of the page for a chapter visited when a single profile is selected. 
    - Implemented changes to show only the original profile selected by the user. 
    - Sent a pull request with the changes. [link : #43](https://github.com/alan-turing-institute/bio-Turing-Way/pull/43) 

### Blockers : 
 None

### Planned :
- Adding color to the badge based on the profile and enabling navigation to the profile page when the badge is clicked.
- Work on handling the condition when the user accesses a chapter not via the pathway's landing page but through the navigation menu or chapter links on the welcome page.

**Comment by arya1302** [comment]
## Status Update (June 28, 2023)

### Work Done :

-   Worked on assigning colours to badges based on the user's profile. However the correct colour of the profile is not being displayed accurately and it needs to be fixed.
-   Worked on enabling navigation to the profile landing page when the badge is clicked. Some style changes of the badge element are required before the work can be pushed.
### Blockers : 
 None

### Planned :
- Complete the task of assigning colours to badges by ensuring the proper mapping of colours to user profiles, and then push the work.
- Complete the implementation of the navigation to the profile landing page when a badge is clicked and push the work. 
- Work on handling the situation where a user accesses a chapter by using the navigation menu or chapter links on the welcome page
- Get suggestions and review from mentors.
- Make additional UI improvements if needed.

**Comment by malvikasharan** [comment]
Thank you so muych @arya1302 for the beautiful work and progressing in your project.
@likeajumprope, @arya1302 and I met today to catch up. She is working on the PR now to fix some issues as described above: https://github.com/alan-turing-institute/bio-Turing-Way/.

Arya, please do share the link of your update notes with us in Slack as GitHub isn't notifying me automatically.
Very excited to see the ideas you are implementing in the test repository.

**Reply by arya1302** [reply]
Thank you @malvikasharan

**Comment by arya1302** [comment]
## Status Update (July 5, 2023)

### Work Done :

-  Completed the implementation for assigning colours to badges. 
-  Fixed the situation where a user accesses a chapter using the navigation menu or links on the welcome page.
-  Started working on the mid-term report. 
### Blockers : 
 None

### Planned :
- Navigation to the profile landing page on badge click still needs to be fixed. So work on that.
- Identify and fix any issues, if present.
- Push all pending changes.
- Get suggestions and review from mentors.
- Continue working on the mid-term report.

**Reply by malvikasharan** [reply]
Brilliant!! Thank you for the update Arya. Posting the link to the preview for others following this discussion: https://deploy-preview-42--the-turing-way-personas.netlify.app/reproducible-research/reproducible-research.html

Pathways currently displayed here but will change when moved over to the main repo: https://deploy-preview-42--the-turing-way-personas.netlify.app/welcome.html#different-pathways

**Comment by arya1302** [comment]
## Status Update (July 12, 2023)

### Work Done :

-  Fixed the issue regarding the navigation to the profile landing page on badge click.
-  Went through the changes and found an issue where, instead of the original profile selected by the user, all badges were displayed when navigating to the next page within a chapter. Started working on this issue.


### Blockers : 
 - A blocker related to pre commit occurred during merging of pull request [#42](https://github.com/alan-turing-institute/bio-Turing-Way/pull/42). The pylint hook failed. 

### Planned :
- Fix the issues found in the implementation.
- Plan and implement the integration of Web Speech API for text to speech conversion.

**Comment by arya1302** [comment]
## Status Update (July 19, 2023)

### Work Done :

-  Worked on the issue where all badges are displayed when navigating to the next page within a chapter, but couldn't complete it. 
-  Worked on implementing the Web Speech API, but there's still much more to be done.
### Blockers : 
 - None
### Planned :
- Resolve the pre commit error occurred during the merging of pull request [42](https://github.com/alan-turing-institute/bio-Turing-Way/pull/42)
- Fix the issue related to the next page navigation within a chapter
- Work on text-to-speech implementation

**Comment by arya1302** [comment]
## Status Update (July 26, 2023)

### Work Done :

-  Worked on implementing the Web Speech API. 
     - Tested it by creating a small website and the API works fine there. However, there are occasional issues with speech synthesis performance. 
     - Implemented the necessary UI components for playing the generated audio in The Turing Way project.
     - Tried to integrate the API into the project, but due to some issues, there is no audio output.
### Blockers : 
 - None
### Planned :
- Work on text-to-speech implementation

**Comment by arya1302** [comment]
## Status Update (August 2 , 2023)

### Work Done :
- The idea described in the PR [#43](https://github.com/alan-turing-institute/bio-Turing-Way/pull/43) does not work well in some situations. So I worked on finding new ways
- Found that sessionStorage can be used to save data for a session. This can be used to keep track of the profile selected by user. Based on this, I started  working on the new approach using sessionStorage.

### Blockers : 
 - None
### Planned :
- Complete the implementation of the new approach and test it.
- Work on moving the implementation to the main repository

**Comment by arya1302** [comment]
## Status Update (August 9 , 2023)

### Work Done :
-  For fixing the issue mentioned in [PR #43](https://github.com/alan-turing-institute/bio-Turing-Way/pull/43), implemented the new approach using sessionStorage for storing user-selected profiles.
- Started working on moving the implementation to the main repository. Went through the [PR](https://github.com/alan-turing-institute/the-turing-way/pull/2246) to get a better idea. 
### Blockers : 
 - None
### Planned :
- Push all pending changes 
- Complete the task of moving the implementation to the main repository.
- Get suggestions and review from mentors.

**Comment by arya1302** [comment]
## Status Update (August 16 , 2023)

### Work Done :
-  Pushed all the pending changes related to [PR #43](https://github.com/alan-turing-institute/bio-Turing-Way/pull/43) and had them reviewed by the mentor.
-  Worked on moving the implementation to the main repository. 
### Blockers : 
 - None
### Planned :
- Complete the task of moving the implementation to the main repository.
- Get suggestions and reviews from mentors.

**Comment by arya1302** [comment]
## Status Update (August 23 , 2023)

### Work Done :
-  Worked on moving the implementation to the main repository.
    - The changes required for the move have been completed and the process is almost done. But i encountered some deployment errors.
    - Sent a pull request with the changes #3277 

### Blockers : 
 - The deployment failed because there were conflicting package versions causing an issue.
### Planned :
- Fix the deployment issue 
- Get suggestions and reviews from mentors.

**Comment by arya1302** [comment]
## Status Update (August 30 , 2023)

### Work Done :
-  Worked on moving the implementation to the main repository.
    -  Resolved errors related to conflicting package versions that I had encountered earlier.
    -  Due to some reason, the netlify deployment has failed again with the error message: 'Failed during stage 'building site': Build script returned a non-zero exit code: 2.
    -  Downgraded the jupyter-book version to 0.11.2 and was able to build the site locally but the netlify deployment still fails
    
### Blockers : 
 - The exact reason behind the netlify deployment failure is unknown
### Planned :
- Fix the netlify deployment error and ensure that the site runs with the implemented changes

**Comment by arya1302** [comment]
## Status Update (September 6 , 2023)

### Work Done :
-  Fixed the Netlify deployment error caused by an incorrect publish directory path.
- Worked on updating the Pathways package implementation to make it compatible with the main project's version, as there was a version mismatch between Jupyter-Book 0.11.2 used by the Pathways package and Jupyter-Book 0.13 used in the main project, which was preventing the card implementation from working correctly in the main project.  This work is not complete, and there is more to be done.
### Blockers : 
None
### Planned :
-  Work on resolving the version compatibility issue between the Pathways package and the main project by updating the Pathways package.
- Get suggestions and reviews from mentors.

**Comment by arya1302** [comment]
## Status Update (September 13 , 2023)

### Work Done :
- Resolved the version compatibility issue between the Pathways package and the main project
    - Added sphinx-panels as an extension, which is necessary to enable proper functionality of welcome page cards.
- Fixed the bug related to the pathway badge displaying "null" when a user clicks on a chapter in the card. 
- Worked on the implementation of text-to-speech feature:
    - Created necessary UI elements, such as a speech button
    - Worked on a basic text to speech implementation using Web Speech API.
### Blockers : 
None
### Planned :
- Complete the implementation of text-to-speech feature.
- Push all the pending changes and get suggestions & reviews from mentors.

**Comment by malvikasharan** [comment]
Brilliant work as usual Arya. Looking forward to working on this over the weekend and merge the pathway feature with a couple of profiles. Very exciting time. 💖
